core/src/main/kotlin/plugins/OpenApi.kt

```diff
@@ -150,3 +159,163 @@ fun Application.configureOpenApi() {
+                typeData = result.typeData,
+                swagger = result.swagger,
+                componentSchemas = result.componentSchemas
+            )
+        }
+    }
+}
+
+/**
+ * An adapted version of [io.github.smiley4.schemakenerator.swagger.steps.SwaggerSchemaCompileReferenceStep] that
+ * generates valid references for the generated schemas (see [getRefPath]).
+ */
+private class CustomCompileReferenceStep {
+    private val schemaUtils = SwaggerSchemaUtils()
+
+    /**
+     * Put referenced schemas into definitions and reference them
+     */
+    fun compile(bundle: Bundle<SwaggerSchema>): CompiledSwaggerSchema {
+        val schemaList = bundle.flatten()
+        val typeDataMap = bundle.buildTypeDataMap()
+        val components = mutableMapOf<String, Schema<*>>()
+
+        val root = resolveReferences(bundle.data.swagger) { refObj ->
+            resolve(refObj, schemaList, typeDataMap, components)
+        }
+
+        return CompiledSwaggerSchema(
+            typeData = bundle.data.typeData,
+            swagger = root,
+            componentSchemas = components
+        )
+    }
+
+    private fun resolve(
+        refObj: Schema<*>,
+        schemaList: List<SwaggerSchema>,
+        typeDataMap: Map<TypeId, BaseTypeData>,
+        components: MutableMap<String, Schema<*>>
+    ): Schema<*> {
+        val referencedId = TypeId.parse(refObj.`$ref`)
+        val referencedSchema = schemaList.find(referencedId)
+        return if (referencedSchema != null) {
+            if (shouldReference(referencedSchema.swagger)) {
+                val refPath = getRefPath(referencedSchema.typeData, typeDataMap)
+                if (!components.containsKey(refPath)) {
+                    components[refPath] = placeholder() // break out of infinite loops
+                    components[refPath] =
+                        resolveReferences(referencedSchema.swagger) { resolve(it, schemaList, typeDataMap, components) }
+                }
+                schemaUtils.referenceSchema(refPath, true)
+            } else {
+                merge(refObj, referencedSchema.swagger)
+            }
+        } else {
+            refObj
+        }
+    }
+
+    private fun placeholder() = Schema<Any>()
+
+    private fun Collection<SwaggerSchema>.find(id: TypeId): SwaggerSchema? {
+        return this.find { it.typeData.id == id }
+    }
+}
+
+/**
+ * An adapted version of [io.github.smiley4.schemakenerator.swagger.steps.SwaggerSchemaCompileUtils.getRefPath] that
+ * generates valid references for the generated schemas. For example, the type `GenericType<Type1, Type2>` is
+ * transformed to `GenericType_Type1-Type2` instead of `GenericType<Type1,Type2>`.
+ *
+ * According to the [specification](https://swagger.io/specification/#components-object), schema keys must match the
+ * regular expression `^[a-zA-Z0-9\.\-_]+$`.
+ */
+private fun getRefPath(typeData: BaseTypeData, typeDataMap: Map<TypeId, BaseTypeData>): String {
```

Comment:
This seems to be mostly the same as `determineTitle()`, can the code be de-duplicated?